### DISCUSSION TITLE
MX09 support (+Android build issues + how I fixed them)

### DISCUSSION THREAD

**Discussion by meglyne** [NaitLee/Cat-Printer#70] [Show and tell]
Hi,

I'm creating this discussion to suggest the support of the MX09 printer to the project, as well as some issues I encountered while building the app for android, how I fixed them, and the process I went through.

## Device issues encountered 

In order of appearance

- [ ] The MX09 device is not supported by the Android app
- [ ] The device is not recognized by default while scanning for devices
- [ ] The device doesn't push paper properly after the print has ended

After installing the project on my Linux programming machine, I've actually been able to print pictures 

In `models.py`, I've added the device name `MX09` to the list (line 21 to 23)

```python
# all known supported models
for name in '_ZZ00 GB01 GB02 GB03 GT01 MX05 MX06 MX08 MX09 YT01'.split(' '):
    Models[name] = Model()
```

After some preliminary testing, I've noticed that the printer was not pushing paper all the way through after printing was completed (the picture was properly printed on paper). This was fixed by adding the model to the list of models with problems feeding paper in `models.py` (line

```python
# that have problem giving feed command
for name in 'MX05 MX06 MX08 MX09'.split(' '):
    Models[name].problem_feeding = True
```

### Build environment

I've used the android build guide to build the app, so I'm running it on an arch linux container in docker, with `mnt/data` mounted on my physical machine.

### Build error n°1 : pyjnius config file missing

This error was caused by a missing file (`config.pxi`) in the pyjnius environment, which determines what the target platform will be.

#### Error message

```console
/root/venv/lib/python3.11/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /root/.local/share/python-for-android/build/other_builds/pyjnius-genericndkbuild/arm64-v8a__ndk_target_21/pyjnius/jnius/jnius.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)

Error compiling Cython file:
------------------------------------------------------------
...
import sys
import traceback

include "jnius_compat.pxi"
include "jni.pxi"
include "config.pxi"
^
------------------------------------------------------------

jnius/jnius.pyx:100:0: 'config.pxi' not found

[...]
```

#### Fix

- Create `config.pxi` in `$DIR_P4A/build/other_builds/pyjnius-genericndkbuild/arm64-v8a__ndk_target_21/pyjnius/jnius`
- Add the following text in the file

```java
DEF JNIUS_PLATFORM = 'android'
```

#### Alternate fix

Run the following command

```bash
echo "DEF JNIUS_PLATFORM = 'android'" > $DIR_P4A/build/other_builds/pyjnius-genericndkbuild/arm64-v8a__ndk_target_21/pyjnius/jnius/config.pxi
```

### Build error n°2 : javac target version unsupported

This error was caused by a misconfiguration in the source and target options of pyjnius. Source and Target options 7 by javac in version 20 of openjdk. Now I'm not sure as of how much it is an issue since I've downgraded to openjdk 17 to fix __Build error n°3__, so take this with a grain of salt and I'll investigate that.

#### Error message

```console
warning: [options] bootstrap class path not set in conjunction with -source 7
error: Source option 7 is no longer supported. Use 8 or later.                                                                                             
error: Target option 7 is no longer supported. Use 8 or later.                                                                                             
Traceback (most recent call last):                                                                                                                         
  File "/mnt/data/@/p4a/build/other_builds/pyjnius-genericndkbuild/arm64-v8a__ndk_target_21/pyjnius/setup.py", line 83, in <module>                        
    compile_native_invocation_handler(JAVA)                                                                                                                
  File "/mnt/data/@/p4a/build/other_builds/pyjnius-genericndkbuild/arm64-v8a__ndk_target_21/pyjnius/setup.py", line 73, in compile_native_invocation_handler                                                                                                                                                          
    subprocess.check_call([                                                                                                                                
  File "/root/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/Lib/subprocess.py", line 369, in check_call               
    raise CalledProcessError(retcode, cmd)                                                                                                                 
subprocess.CalledProcessError: Command '['/usr/lib/jvm/java-20-openjdk/bin/javac', '-target', '1.7', '-source', '1.7', 'jnius/src/org/jnius/NativeInvocationHandler.java']' returned non-zero exit status 2. 
```

#### Fix 

I opened the file `$DIR_P4A/build/other_builds/pyjnius-genericndkbuild/arm64-v8a__ndk_target_21/pyjnius/setup.py`and changed `source_level` from `'1.7'` to `'1.8'`

```python
def compile_native_invocation_handler(java):
    '''Find javac and compile NativeInvocationHandler.java.'''
    javac = java.get_javac()
    source_level = '1.8' #'1.7'
```

### Build error n°3 : class file major version unsupported by Gradle

This error happens because Gradle doesn't support `class file major version 64` while building `build.gradle`. I've fixed it by doing two things:
- Upgrading the gradle version in `gradle-wrapper.properties` from 7.4.1 to 7.6.2 (this changed the format of the error message to
```console
A problem occurred configuring root project 'cat-printer'.
> Could not open cp_proj generic class cache for build file '/mnt/data/@/p4a/dists/cat-printer/build.gradle' (/root/.gradle/caches/7.6.2/scripts/ble0a2x2p2fwmr6a9wi3dfn3e).
   > BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 64
```
but the error remained the same.
- Downgrading the openjdk version from 20 to 17

#### Error message

```console
Welcome to Gradle 7.4.1!

Here are the highlights of this release:
 - Aggregated test and JaCoCo reports
 - Marking additional test source directories as tests in IntelliJ
 - Support for Adoptium JDKs in Java toolchains

For more details see https://docs.gradle.org/7.4.1/release-notes.html

Starting a Gradle Daemon (subsequent builds will be faster)

FAILURE: Build failed with an exception.

* Where:
Build file '/mnt/data/@/p4a/dists/cat-printer/build.gradle'

* What went wrong:
Could not compile build file '/mnt/data/@/p4a/dists/cat-printer/build.gradle'.
> startup failed:
  General error during conversion: Unsupported class file major version 64
  
  java.lang.IllegalArgumentException: Unsupported class file major version 64
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:199)
        [...]
```

#### Fix part 1 : Upgrading gradle

In `$DIR_P4A/dists/cat-printer/gradle/wrapper/gradle-wrapper.properties`, I've changed the Url of the Gradle downloaded version from version `7.4.1`to version `7.6.2` 
```
[...]
distributionUrl=https\://services.gradle.org/distributions/gradle-7.4.1-all.zip
```
To :
```
[...]
distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.2-all.zip
```

#### Fix part 2 : Downgrading openjdk

The new error message led me to the following [stackoverflow issue](https://stackoverflow.com/questions/75651945/bug-exception-in-phase-semantic-analysis-in-source-unit-buildscript-unsup). According to the solution, this issue may also have been caused because of incompatibilites in support between gradle and openjdk. The default installation of openjsk with `pacman -Syuu jdk-openjdk` installs openjdk version 20. So I followed the [Archlinux wiki article on java](https://wiki.archlinux.org/title/java) to install openjdk version 17 and to enable it by default like so

```console
(env) [root@97a8c2f2d8f6 cat-printer]# pacman -Syuu jdk17-openjdk
(env) [root@97a8c2f2d8f6 cat-printer]# archlinux-java set java-17-openjdk
(env) [root@97a8c2f2d8f6 cat-printer]# archlinux-java status
```
Showing openjdk17 as the default installation

#### Remarks

I'm not sure how much upgrading Gradle to version 7.6.2 has fixed the issue and I'd say it's mostly using openjdk 17. So I'll try to test further on that.

### Additional python package install remarks

On the arch linux docker install, running pip while not sourcing a virtualenv triggers the following error message.

``` console
[root@97a8c2f2d8f6 ~]# pip install cython
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try 'pacman -S
    python-xyz', where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Arch-packaged Python package,
    create a virtual environment using 'python -m venv path/to/venv'.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip.
    
    If you wish to install a non-Arch packaged Python application,
    it may be easiest to use 'pipx install xyz', which will manage a
    virtual environment for you. Make sure you have python-pipx
    installed via pacman.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

While I understand this is running in a docker container created specifically for building the android app, I believe using a virtual environment is the best practice there. Therefore, I have installed the python packages like so 

```console
[root@97a8c2f2d8f6 /]# cd ~
[root@97a8c2f2d8f6 ~]# python -m venv env/
[root@97a8c2f2d8f6 ~]# source env/bin/activate
(env) [root@97a8c2f2d8f6 ~]# pip install python-for-android cython bleak
```

## Conclusion

While I'm not sure how to fix the build errors regarding pyjnius from the source code, I believe specifying the openjdk version while installing the arch linux packages in the docker container will solve a lot of further issues if your intent is to make minimal changes to the android build instruction manual. I've always had a lot of trouble with building apps with gradle for android or java applications and it's always a mess in my opinion, but I'm glad I made it in the end. In that regard, the android building guide was very straight-forward and helped me a lot! 

## Context
I've bought recently a cat printer from Amazon. I've tried to use it on Android through the bloated app "Fun Print" and I've been taken aback by the amount of permissions requested and the fact that it was annoyingly turning the Bluetooth back on every time I was turning it of (concerning isn't it ?). I've stumbled upon the Cat-Printer project and I've installed the apk, but I haven't been able to use it because the name MX09 wasn't recognized by the app, and that's what made me contribute. While adding the device to the list of supported devices was pretty straight-forward, the building process for android took me a lot of time because of the errors that I've mentioned.  

It's my first time contributing to an open-source project and I thought I'd share my experience. I hope I've not been too verbose...

**Comment by NaitLee** [comment]
Cheers for your achieving experience 🎉

A fun fact about this project is that who come here are either pros or wanna-be geeks that have made their first step.

I just had free time, and would like to make a new release including community contributions like this one.  
Your experience on building may help a lot…

I’m glad this project have given ones feeling of achievement… wonderful!

Tip: you may want to submit pyjnius related issues to [its dedicated repo](https://github.com/kivy/pyjnius). I’m not experienced with Java or Android or so, and have no relationship with it (other than a common user)
